### PR TITLE
[camera] Fix the orientation of videos recorded in landscape on Android

### PR DESCRIPTION
**Pull request by navaronbracke**
This PR fixes a bug with the orientation of the recorded video on Android. In Landscape Left and Landscape Right the video was flipped 180 degrees in the wrong direction. I also added a new test for the fallback-to-sensor-orientation case for portrait. Only testing that case against landscape felt a little odd to me.

Fixes https://github.com/flutter/flutter/issues/27201, https://github.com/flutter/flutter/issues/98486

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

<details>
  <summary>Before / After samples</summary>
  
 ### Landscape Left Recording (rotated the device 90CCW, then started recording)

 #### Before

https://user-images.githubusercontent.com/23480364/155849185-abb6080e-eb14-4145-a2d6-3cab3a6a8310.mp4

 #### After 

https://user-images.githubusercontent.com/23480364/155849236-d8218434-4354-4e23-ad3a-5ff842c86642.mp4

 ### LandscapeRight Recording (rotated the device 90CW, then started recording)

 #### Before

https://user-images.githubusercontent.com/23480364/155849199-29d65b4b-fcc2-4457-aa25-050a6e54789b.mp4

 #### After 

https://user-images.githubusercontent.com/23480364/155849268-54c5d7f8-ee70-4995-be62-ad2d092ae002.mp4

</details>

### CODE REVIEW

File: packages/camera/camera/CHANGELOG.md
Comment (navaronbracke) [reply packages/camera/camera/CHANGELOG.md]: Fixed it and the check is passing now.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
There's nothing here that explains why this code is right and the previous code was wrong; the documentation of the method doesn't provide any context that allows evaluating it. A PR that reverted this change would look exactly the same as this one, and would be similarly unclear.

Please update the documentation of this method to explain what exactly the values are meant to express in a way that allows clearly evaluating whether `90` or `270` is right. (Ideally that means with links to relevant documentation.)

**Comment by navaronbracke** [comment]
> There's nothing here that explains why this code is right and the previous code was wrong; the documentation of the method doesn't provide any context that allows evaluating it. A PR that reverted this change would look exactly the same as this one, and would be similarly unclear.
> 
> Please update the documentation of this method to explain what exactly the values are meant to express in a way that allows clearly evaluating whether `90` or `270` is right. (Ideally that means with links to relevant documentation.)

I'll add a link to https://developer.android.com/reference/android/media/MediaRecorder#setOrientationHint(int)
In the documentation for that method, the `degrees` argument is explained as follows
```
the angle to be rotated clockwise in degrees.
```

**Comment by navaronbracke** [comment]
@stuartmorgan Updated the method's documentation with a link that illustrates the calculation.

**Comment by navaronbracke** [comment]
I rebased to fix the changelog and pubspec. We'd probably have to wait until the other changes land first?

**Comment by navaronbracke** [comment]
@stuartmorgan should be good now. I went ahead and reset to before the bad rebase, then did a merge of main and finally I added the missing changes from after the bad rebase.

**Comment by ditman** [comment]
(Removing myself as a reviewer, I think I got added as an owner during a bad rebase)

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java
Comment (stuartmorgan) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:163]: The linked documentation specifically discusses both clockwise and counterclockwise rotation, so it doesn't really clarify the thing I was asking about. This comment should say whether the rotation returned is clockwise or counterclockwise, and why.

In fact, based on that documentation it's not clear to me that this method is even intended to do. Is it supposed to apply to front-facing or back-facing cameras? Or is it both, in which case this PR is as wrong as the original code, just for the opposite cameras.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java
Comment (navaronbracke) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:166]: According to the docs for `MediaRecorder.setOrientationHint(int)` the rotation is in clockwise degrees. That documentation does not specifically mention a front or back facing camera, so I assume it refers to both? That paragraph is rather slim when it comes to actual information, but this method is the only consumer of the computation, so I don't really see what else could be of use in terms of docs?

Comment (navaronbracke) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:166]: > Or is it both, in which case this PR is as wrong as the original code, just for the opposite cameras.

I tested the fix on my phone with both front & back camera's in both landscape left and right. All situations did produce the correct results. I could verify the fix on a tablet as well, just to be sure?

File: packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManagerTest.java
Comment (navaronbracke) [inline packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManagerTest.java:71]: Not sure if having separate tests for `getVideoOrientationBackFacing_xxx` and `getVideoOrientationFrontFacing_xxx` is a good thing? Back/front facing seems to not have an effect on the switch case that was changed.

**Comment by Code-Remote** [comment]
Ah it looks like the fix is already on the way, I was just planning to ask around. I really thought I checked all the threads carefully, but looks like this one is quite recent. Keep me up to date about the progress, or if there is anything I can do please let me know. (https://github.com/flutter/flutter/issues/98486#issuecomment-1088648739)

**Comment by Code-Remote** [comment]
Any update on this pull request, or anything I can help with? Looking forward to use it 👍

File: packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManagerTest.java
Comment (navaronbracke) [reply packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManagerTest.java]: Per an offline discussing with @Code-Remote I removed the front facing tests as those are redundant.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java
Comment (navaronbracke) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:145]: I'm not sure if I should mention that this is for both front and back-facing cameras, because the documentation I  referred to didn't specifically mention that. I'd rather only document something of which there is actual proof that can be backed up by the Android docs.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java
Comment (stuartmorgan) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/sensororientation/DeviceOrientationManager.java:166]: Resolving since the first of the two links explains that it's clockwise; I missed that before. I am still confused as to why front-facing isn't opposite, but maybe it's specifically to avoid this kind of confusion.

File: packages/camera/camera/CHANGELOG.md
Comment (stuartmorgan) [inline packages/camera/camera/CHANGELOG.md:3]: Nit: "videos", not "video's"

**Review by stuartmorgan — APPROVED** [review]
LGTM with nit, sorry for the delay in getting back to this.

+ @camsim99 for second review.

File: packages/camera/camera/CHANGELOG.md
Comment (navaronbracke) [reply packages/camera/camera/CHANGELOG.md]: Done

**Review by camsim99 — APPROVED** [review]
LGTM!

**Comment by stuartmorgan** [comment]
Should land shortly. Thanks for the contribution, and for the investigation into the relevant docs!

**Pull request merged**
